src/Core/AdminConsole/OrganizationFeatures/Policies/PolicyValidators/SingleOrgPolicyValidator.cs

```diff
@@ -80,17 +85,42 @@ private async Task RemoveNonCompliantUsersAsync(Guid organizationId)
 
         var userOrgs = await _organizationUserRepository.GetManyByManyUsersAsync(
                 removableOrgUsers.Select(ou => ou.UserId!.Value));
-        foreach (var orgUser in removableOrgUsers)
+
+        if (_featureService.IsEnabled(FeatureFlagKeys.AccountDeprovisioning))
         {
-            if (userOrgs.Any(ou => ou.UserId == orgUser.UserId
-                        && ou.OrganizationId != org.Id
-                        && ou.Status != OrganizationUserStatusType.Invited))
+            var isOwner = await _currentContext.OrganizationOwner(organizationId);
+
+            var revocableUsers = removableOrgUsers.Where(nonCompliantUser => orgUsers.Any(orgUser =>
+                nonCompliantUser.UserId == orgUser.UserId
+                && nonCompliantUser.OrganizationId != org.Id
+                && nonCompliantUser.Status != OrganizationUserStatusType.Invited))
+                .ToList();
+
+            var commandResult = await _revokeNonCompliantOrganizationUserCommand.RevokeNonCompliantOrganizationUsersAsync(
+                new RevokeOrganizationUsersRequest(organizationId, revocableUsers, new StandardUser(savingUserId ?? Guid.Empty, isOwner)));
+
+            if (commandResult.HasErrors)
             {
-                await _removeOrganizationUserCommand.RemoveUserAsync(organizationId, orgUser.Id,
-                    savingUserId);
+                throw new BadRequestException(string.Join(", ", commandResult.ErrorMessages));
+            }
 
-                await _mailService.SendOrganizationUserRemovedForPolicySingleOrgEmailAsync(
-                    org.DisplayName(), orgUser.Email);
+            await Task.WhenAll(revocableUsers.Select(x =>
+                _mailService.SendOrganizationUserRevokedForPolicySingleOrgEmailAsync(org.DisplayName(), x.Email)));
+        }
```

Comment:
The command does not allow a mixed success / fail state, right? I was initially concerned that if any member could not be revoked, then nobody gets their email. But looking at the command, it validates everyone upfront and then doesn't proceed if there are any errors, so I think that's OK.

That said, if an admin revokes multiple users from the UI, we _do_ allow a success or failure state per user, so we may need to accommodate both behaviors if the command is expanded to replace the OrganizationService methods.

Reply:
The complicated thing about this command is all the validation logic.  What we could do is break out the validation into a validator class, then have two commands (one for the 2FA/Single Org policy enablement and one for a user explicitly revoking a group of users). Then each command can make the decision about enforcing the validation result.

Unless we just want the command to just only reject invalid responses. Then we would make a Response object to extend CommandResult.  Maybe something like
```
public class RevokeOrganizationUserCommandResponse : CommandResult 
{
...
public Guid[] SuccessfullyRevokedUsers { get; } 
public Guid[] FailedToRevokeUsers { get; }
...
}

```

That way we should show the partial success of the action.

Reply:
All good ideas. I've asked Priya for feedback in the team channel about whether we can align this behavior. No other changes required for now.